### ISSUE TITLE
Disk metrics not captured in some environments

### ISSUE THREAD

**Issue by askfongjojo** [oxidecomputer/crucible#942]
We've noticed that disk metrics are not shown on rack3, e.g.
https://$oxide-silo-rack3-endpoint/projects/demo/instances/docker-engine/metrics
And I saw that again when a customer was screen-sharing a VM that had disk r/w load test going on. I tried troubleshooting a bit on rack3. There was no error in the Nexus log - the metrics API calls all responded HTTP 200. When I ran those API calls directly, no datapoint was returned. Here is an example:
```
oxide api '/v1/disks/docker-engine/metrics/read_bytes?project=demo&start_time=2023-09-12T16%3A34%3A22.508Z&end_time=2023-09-15T16%3A34%3A22.508Z&limit=30'
{
  "items": [],
  "next_page": null
}
```

This issue was not seen on rack2 which currently has a newer build but I am rather certain that the issue was not seen back when it was on the same version as rack3 (release 1.0.2, i.e. omicron commit f2bc296f8d6942bf6b86fafeb18fdc52e50a57b7).

@leftwo - Let me know if there are certain log files I should grab from rack3 to further debug this (I tried looking at crucible logs but haven't managed to pinpoint any error).

**Comment by leftwo** [comment]
The only idea I have for logs would be to capture the propolis logs for the start of an instance.
During startup, there is a message printed when the upstairs registers with Oximeter, and that might 
provide some clues.

**Comment by bnaecker** [comment]
It may help to copy the `oximeter` log files as well. That will show us if any collections at all were made, or when `oximeter` first encountered an error.

**Comment by leftwo** [comment]
The upstairs should print something like this:

```
{"msg":"Crucible stats registered with UUID: b0b70141-1df9-44aa-83ef-b51719947edc","v":0,"name":"crucible","level":30,"time":"2023-09-23T21:21:11.509896-07:00","hostname":"local.local","pid":10613}
```

**Comment by leftwo** [comment]
If there is an error, you might see this message:
`error!(up_oxc.log, "Failed to register metric producer: {}", e);`

**Comment by askfongjojo** [comment]
@leftwo / @bnaecker: Thank you for the pointers.

Oximeter producer registrstion of the propolis zone was successful:
```
16:16:06.758Z INFO propolis-server: Metrics server will use MetricsEndpointConfig { propolis_addr: [fd00:1122:3344:112::22]:12400, metric_addr: [fd00:1122:3344:108::3]:12221 }
16:16:06.759Z INFO propolis-server: Starting server...
16:16:06.760Z INFO propolis-server: listening
    local_addr = [fd00:1122:3344:112::22]:12400
16:16:06.968Z INFO propolis-server: accepted connection
    local_addr = [fd00:1122:3344:112::22]:12400
    remote_addr = [fd00:1122:3344:112::1]:56389
16:16:06.968Z INFO propolis-server: request completed
    error_message_external = Not Found
    error_message_internal = Server not initialized (no instance)
    local_addr = [fd00:1122:3344:112::22]:12400
    method = GET
    remote_addr = [fd00:1122:3344:112::1]:56389
    req_id = 267b375e-2ad4-485a-a3a3-30abb410b2ab
    response_code = 404
    uri = /instance
16:16:07.005Z INFO propolis-server: Attempt to register [fd00:1122:3344:112::22]:0 with Nexus/Oximeter at [fd00:1122:3344:108::3]:12221
    local_addr = [fd00:1122:3344:112::22]:12400
    method = PUT
    remote_addr = [fd00:1122:3344:112::1]:56389
    req_id = a89d7567-d377-4f76-b038-05b01930ae75
    uri = /instance
Sep 14 16:16:07.006 INFO listening, local_addr: [fd00:1122:3344:112::22]:63024, component: dropshot
Sep 14 16:16:07.107 INFO starting oximeter metric server, address: [fd00:1122:3344:112::22]:63024, producer_id: c07ceb29-ea72-42a7-bde0-0851a8084824, route: /collect/c07ceb29-ea72-42a7-bde0-0851a8084824
16:16:07.107Z INFO propolis-server: connected [fd00:1122:3344:112::22]:0 to oximeter [fd00:1122:3344:108::3]:12221
    local_addr = [fd00:1122:3344:112::22]:12400
    method = PUT
    remote_addr = [fd00:1122:3344:112::1]:56389
    req_id = a89d7567-d377-4f76-b038-05b01930ae75
    uri = /instance
```
It was able to communicate with oximeter: 
```
root@oxz_propolis-server_e6807cfe-018f-4ff0-95a6-4dde0ff3322e:~# grep c07ceb29-ea72-42a7-bde0-0851a8084824 /var/svc/log/system-illumos-propolis-server\:default.log
...
Sep 24 18:27:47.215 INFO request completed, response_code: 200, uri: /collect/c07ceb29-ea72-42a7-bde0-0851a8084824, method: GET, req_id: dac58508-412c-4841-9dbc-77e1f8a9d9a9, remote_addr: [fd00:1122:3344:10b::3]:48218, local_addr: [fd00:1122:3344:112::22]:63024, component: dropshot
Sep 24 18:28:17.216 INFO request completed, response_code: 200, uri: /collect/c07ceb29-ea72-42a7-bde0-0851a8084824, method: GET, req_id: 707944b9-9e1e-4c1e-a01d-19e6977159cf, remote_addr: [fd00:1122:3344:10b::3]:48218, local_addr: [fd00:1122:3344:112::22]:63024, component: dropshot
```

However, oximeter has been throwing the following error when it attempts to collect metrics:
```
18:31:17.223Z INFO oximeter (oximeter-agent): collecting from producer
    collector_id = e9826cdc-6d3a-4eff-b1b5-ec4364ebe6b9
    file = oximeter/collector/src/lib.rs:75
    producer_id = c07ceb29-ea72-42a7-bde0-0851a8084824
18:31:17.224Z WARN oximeter (oximeter-agent): failed to collect results from producer: error decoding response body: invalid type: sequence, expected a map at line 1 column 268
    collector_id = e9826cdc-6d3a-4eff-b1b5-ec4364ebe6b9
    file = oximeter/collector/src/lib.rs:97
    producer_id = c07ceb29-ea72-42a7-bde0-0851a8084824
```

**Comment by askfongjojo** [comment]
FWIW, utilization metrics (i.e. provisioned cpu/memory/disk) are working just fine so it's likely not a problem with oximeter.  The most recent changes in `oximeter/collector/src/lib.rs` also do not appear to have changed what the producer request payload should look like.

**Comment by bnaecker** [comment]
I'm pretty sure this is https://github.com/oxidecomputer/omicron/pull/3878 coming up again. I changed the way fields are listed, from an array to a map. That was suggested during code review, and I didn't think carefully enough about the implications for clients. I expect that an update will fix both sides, but we can verify by looking at the two sides. If the version of `oximeter` on which crucible depends is before `8ffd476f8af7a590752265d5e8564d8bc1534eee`, but the commit of `omicron` itself is later, then this will be an issue.

**Comment by askfongjojo** [comment]
https://github.com/oxidecomputer/crucible/commit/7ba88771215818d053bb64ae9e995f41ffb3c5a7 caught up on the oximeter change before we released v1.0.2 but crucible version wasn't bumped in Omicron after that (not for 2 weeks but we never noticed, until https://github.com/oxidecomputer/omicron/commit/25c73f4c9c360ba80083e0bd01d435ca21ec471c). So that explains why the issue happens only for the TUF artifacts that has https://github.com/oxidecomputer/omicron/pull/3878 **and** an older crucible:
```
16:16:07.205Z INFO propolis-server (crucible-620f7e27-a55e-4a44-a97a-0b75cc3f35c4): Crucible Version: BuildInfo {
        version: "0.0.1",
        git_sha: "5dc77d87dec8bbf56a603821d67ad13f47f99f95",
        git_commit_timestamp: "2023-08-25T01:42:07.000000000Z",
        git_branch: "master",
        rustc_semver: "1.72.0",
        rustc_channel: "stable",
        rustc_host_triple: "x86_64-unknown-illumos",
        rustc_commit_sha: "5680fa18feaa87f3ff04063800aec256c3d4b4be",
        cargo_triple: "x86_64-unknown-illumos",
        debug: false,
        opt_level: 3,
    }
```
